### PR TITLE
Refine pre_reorder_triton_kernel slightly to improve performance

### PR DESCRIPTION
**Pull request by yuan-luo**
## Motivation

In ep_moe kernel _pre_reorder_triton_kernel_ and _post_reorder_triton_kernel_, every inner loop recomputes
offset = start_offset + tl.arange(...)

The optimization is to create a constant once:
vec = tl.arange(0, BLOCK_SIZE)
and inside the loop use idx = start_offset + vec

The benefit is one less instruction each iteration. The warp scheduler can vectorize the access pattern.
Per benchmark result the kernel gains 10-15% performance.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

**Comment by yuan-luo** [comment]
The benchmark script:

```
# test_pre_reorder_triton_kernel.py

import torch
import triton
import triton.language as tl
import time

@triton.jit
def pre_reorder_triton_kernel(
    input_ptr,
    gateup_input_ptr,
    src2dst_ptr,
    topk_ids_ptr,
    a1_scales_ptr,
    start_expert_id,
    end_expert_id,
    topk,
    hidden_size,
    BLOCK_SIZE: tl.constexpr,
):
    OutDtype = gateup_input_ptr.dtype.element_ty

    src_idx = tl.program_id(0)
    src2dst_ptr = src2dst_ptr + src_idx * topk
    topk_ids_ptr = topk_ids_ptr + src_idx * topk
    src_ptr = input_ptr + src_idx * hidden_size

    vec = tl.arange(0, BLOCK_SIZE)

    for idx in range(topk):
        expert_id = tl.load(topk_ids_ptr + idx)
        if expert_id >= start_expert_id and expert_id <= end_expert_id:
            if a1_scales_ptr is not None:
                scale = 1.0 / tl.load(a1_scales_ptr + expert_id - start_expert_id)
            else:
                scale = 1.0

            dst_idx = tl.load(src2dst_ptr + idx)
            dst_ptr = gateup_input_ptr + dst_idx * hidden_size
            for start_offset in tl.range(0, hidden_size, BLOCK_SIZE):
                offset = start_offset + vec
                mask = offset < hidden_size
                in_data = tl.load(src_ptr + offset, mask=mask).to(tl.float32)
                out_data = (in_data * scale).to(OutDtype)
                tl.store(dst_ptr + offset, out_data, mask=mask)

def main():
    BATCH = 640
    TOPK = 2
    HIDDEN_SIZE = 1024
    BLOCK_SIZE = 512
    EXPERT_RANGE = (0, 255)

    input_ptr = torch.randn(BATCH, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
    gateup_input_ptr = torch.zeros(BATCH * TOPK, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
    src2dst_ptr = torch.randint(0, BATCH * TOPK, (BATCH, TOPK), dtype=torch.int32, device='cuda')
    topk_ids_ptr = torch.randint(EXPERT_RANGE[0], EXPERT_RANGE[1] + 1, (BATCH, TOPK), dtype=torch.int32, device='cuda')
    a1_scales_ptr = torch.rand(EXPERT_RANGE[1] - EXPERT_RANGE[0] + 1, dtype=torch.float32, device='cuda')

    input_ptr = input_ptr.view(-1)
    gateup_input_ptr = gateup_input_ptr.view(-1)
    src2dst_ptr = src2dst_ptr.view(-1)
    topk_ids_ptr = topk_ids_ptr.view(-1)

    def run_kernel():
        pre_reorder_triton_kernel_fast[(BATCH,)](
            input_ptr,
            gateup_input_ptr,
            src2dst_ptr,
            topk_ids_ptr,
            a1_scales_ptr,
            EXPERT_RANGE[0],
            EXPERT_RANGE[1],
            TOPK,
            HIDDEN_SIZE,
            BLOCK_SIZE
        )

    for _ in range(10):
        run_kernel()
    torch.cuda.synchronize()

    # Benchmark
    n_iter = 100000
    start = time.time()
    for _ in range(n_iter):
        run_kernel()
    torch.cuda.synchronize()
    end = time.time()

    print(f"Average execution time over {n_iter} iterations: {(end - start) / n_iter * 1000:.3f} ms")

if __name__ == "__main__":
    main()
```

Before the fix result:
```
[root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
Average execution time over 100000 iterations: 0.026 ms
```
After fix result:
```
[root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
Average execution time over 100000 iterations: 0.024 ms
```

**Comment by BBuf** [comment]
> The benchmark script:
> 
> ```
> # test_pre_reorder_triton_kernel.py
> 
> import torch
> import triton
> import triton.language as tl
> import time
> 
> @triton.jit
> def pre_reorder_triton_kernel(
>     input_ptr,
>     gateup_input_ptr,
>     src2dst_ptr,
>     topk_ids_ptr,
>     a1_scales_ptr,
>     start_expert_id,
>     end_expert_id,
>     topk,
>     hidden_size,
>     BLOCK_SIZE: tl.constexpr,
> ):
>     OutDtype = gateup_input_ptr.dtype.element_ty
> 
>     src_idx = tl.program_id(0)
>     src2dst_ptr = src2dst_ptr + src_idx * topk
>     topk_ids_ptr = topk_ids_ptr + src_idx * topk
>     src_ptr = input_ptr + src_idx * hidden_size
> 
>     vec = tl.arange(0, BLOCK_SIZE)
> 
>     for idx in range(topk):
>         expert_id = tl.load(topk_ids_ptr + idx)
>         if expert_id >= start_expert_id and expert_id <= end_expert_id:
>             if a1_scales_ptr is not None:
>                 scale = 1.0 / tl.load(a1_scales_ptr + expert_id - start_expert_id)
>             else:
>                 scale = 1.0
> 
>             dst_idx = tl.load(src2dst_ptr + idx)
>             dst_ptr = gateup_input_ptr + dst_idx * hidden_size
>             for start_offset in tl.range(0, hidden_size, BLOCK_SIZE):
>                 offset = start_offset + vec
>                 mask = offset < hidden_size
>                 in_data = tl.load(src_ptr + offset, mask=mask).to(tl.float32)
>                 out_data = (in_data * scale).to(OutDtype)
>                 tl.store(dst_ptr + offset, out_data, mask=mask)
> 
> def main():
>     BATCH = 640
>     TOPK = 2
>     HIDDEN_SIZE = 1024
>     BLOCK_SIZE = 512
>     EXPERT_RANGE = (0, 255)
> 
>     input_ptr = torch.randn(BATCH, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
>     gateup_input_ptr = torch.zeros(BATCH * TOPK, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
>     src2dst_ptr = torch.randint(0, BATCH * TOPK, (BATCH, TOPK), dtype=torch.int32, device='cuda')
>     topk_ids_ptr = torch.randint(EXPERT_RANGE[0], EXPERT_RANGE[1] + 1, (BATCH, TOPK), dtype=torch.int32, device='cuda')
>     a1_scales_ptr = torch.rand(EXPERT_RANGE[1] - EXPERT_RANGE[0] + 1, dtype=torch.float32, device='cuda')
> 
>     input_ptr = input_ptr.view(-1)
>     gateup_input_ptr = gateup_input_ptr.view(-1)
>     src2dst_ptr = src2dst_ptr.view(-1)
>     topk_ids_ptr = topk_ids_ptr.view(-1)
> 
>     def run_kernel():
>         pre_reorder_triton_kernel_fast[(BATCH,)](
>             input_ptr,
>             gateup_input_ptr,
>             src2dst_ptr,
>             topk_ids_ptr,
>             a1_scales_ptr,
>             EXPERT_RANGE[0],
>             EXPERT_RANGE[1],
>             TOPK,
>             HIDDEN_SIZE,
>             BLOCK_SIZE
>         )
> 
>     for _ in range(10):
>         run_kernel()
>     torch.cuda.synchronize()
> 
>     # Benchmark
>     n_iter = 100000
>     start = time.time()
>     for _ in range(n_iter):
>         run_kernel()
>     torch.cuda.synchronize()
>     end = time.time()
> 
>     print(f"Average execution time over {n_iter} iterations: {(end - start) / n_iter * 1000:.3f} ms")
> 
> if __name__ == "__main__":
>     main()
> ```
> 
> Before the fix result:
> 
> ```
> [root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
> Average execution time over 100000 iterations: 0.026 ms
> ```
> 
> After fix result:
> 
> ```
> [root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
> Average execution time over 100000 iterations: 0.024 ms
> ```

Good job. Can you add the benchmark script to https://github.com/sgl-project/sglang/tree/main/benchmark/kernels/fused_moe_triton?

**Comment by yuan-luo** [comment]
> > The benchmark script:
> > ```
> > # test_pre_reorder_triton_kernel.py
> > 
> > import torch
> > import triton
> > import triton.language as tl
> > import time
> > 
> > @triton.jit
> > def pre_reorder_triton_kernel(
> >     input_ptr,
> >     gateup_input_ptr,
> >     src2dst_ptr,
> >     topk_ids_ptr,
> >     a1_scales_ptr,
> >     start_expert_id,
> >     end_expert_id,
> >     topk,
> >     hidden_size,
> >     BLOCK_SIZE: tl.constexpr,
> > ):
> >     OutDtype = gateup_input_ptr.dtype.element_ty
> > 
> >     src_idx = tl.program_id(0)
> >     src2dst_ptr = src2dst_ptr + src_idx * topk
> >     topk_ids_ptr = topk_ids_ptr + src_idx * topk
> >     src_ptr = input_ptr + src_idx * hidden_size
> > 
> >     vec = tl.arange(0, BLOCK_SIZE)
> > 
> >     for idx in range(topk):
> >         expert_id = tl.load(topk_ids_ptr + idx)
> >         if expert_id >= start_expert_id and expert_id <= end_expert_id:
> >             if a1_scales_ptr is not None:
> >                 scale = 1.0 / tl.load(a1_scales_ptr + expert_id - start_expert_id)
> >             else:
> >                 scale = 1.0
> > 
> >             dst_idx = tl.load(src2dst_ptr + idx)
> >             dst_ptr = gateup_input_ptr + dst_idx * hidden_size
> >             for start_offset in tl.range(0, hidden_size, BLOCK_SIZE):
> >                 offset = start_offset + vec
> >                 mask = offset < hidden_size
> >                 in_data = tl.load(src_ptr + offset, mask=mask).to(tl.float32)
> >                 out_data = (in_data * scale).to(OutDtype)
> >                 tl.store(dst_ptr + offset, out_data, mask=mask)
> > 
> > def main():
> >     BATCH = 640
> >     TOPK = 2
> >     HIDDEN_SIZE = 1024
> >     BLOCK_SIZE = 512
> >     EXPERT_RANGE = (0, 255)
> > 
> >     input_ptr = torch.randn(BATCH, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
> >     gateup_input_ptr = torch.zeros(BATCH * TOPK, HIDDEN_SIZE, dtype=torch.float16, device='cuda')
> >     src2dst_ptr = torch.randint(0, BATCH * TOPK, (BATCH, TOPK), dtype=torch.int32, device='cuda')
> >     topk_ids_ptr = torch.randint(EXPERT_RANGE[0], EXPERT_RANGE[1] + 1, (BATCH, TOPK), dtype=torch.int32, device='cuda')
> >     a1_scales_ptr = torch.rand(EXPERT_RANGE[1] - EXPERT_RANGE[0] + 1, dtype=torch.float32, device='cuda')
> > 
> >     input_ptr = input_ptr.view(-1)
> >     gateup_input_ptr = gateup_input_ptr.view(-1)
> >     src2dst_ptr = src2dst_ptr.view(-1)
> >     topk_ids_ptr = topk_ids_ptr.view(-1)
> > 
> >     def run_kernel():
> >         pre_reorder_triton_kernel_fast[(BATCH,)](
> >             input_ptr,
> >             gateup_input_ptr,
> >             src2dst_ptr,
> >             topk_ids_ptr,
> >             a1_scales_ptr,
> >             EXPERT_RANGE[0],
> >             EXPERT_RANGE[1],
> >             TOPK,
> >             HIDDEN_SIZE,
> >             BLOCK_SIZE
> >         )
> > 
> >     for _ in range(10):
> >         run_kernel()
> >     torch.cuda.synchronize()
> > 
> >     # Benchmark
> >     n_iter = 100000
> >     start = time.time()
> >     for _ in range(n_iter):
> >         run_kernel()
> >     torch.cuda.synchronize()
> >     end = time.time()
> > 
> >     print(f"Average execution time over {n_iter} iterations: {(end - start) / n_iter * 1000:.3f} ms")
> > 
> > if __name__ == "__main__":
> >     main()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Before the fix result:
> > ```
> > [root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
> > Average execution time over 100000 iterations: 0.026 ms
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > After fix result:
> > ```
> > [root@decfee1df170 pre_reorder]# python test_pre_reorder_triton_kernel.py 
> > Average execution time over 100000 iterations: 0.024 ms
> > ```
> 
> Good job. Can you add the benchmark script to https://github.com/sgl-project/sglang/tree/main/benchmark/kernels/fused_moe_triton?

Hi @BBuf, thanks for the comments. I've added benchmark test case in the folder.

**Comment by yuan-luo** [comment]
```
[root@decfee1df170 fused_moe_triton]# python benchmark_pre_reorder_triton.py 
INFO 05-27 10:42:57 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-27 10:42:59 [__init__.py:239] Automatically detected platform cuda.
Average execution time over 100000 iterations: 0.025 ms
```

```
[root@decfee1df170 fused_moe_triton]# nvidia-smi 
Tue May 27 10:43:11 2025       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.06             Driver Version: 535.183.06   CUDA Version: 12.4     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA H20                     On  | 00000000:08:00.0 Off |                    0 |
| N/A   39C    P0             143W / 500W |   1695MiB / 97871MiB |    100%      Default |
|                                         |                      |             Disabled |
......
```

**Comment by yuan-luo** [comment]
Refactored the test case according to @BBuf 's comments.

Before fix, the result is:
```
[root@decfee1df170 fused_moe_triton]# python benchmark_pre_reorder_triton.py --hidden-size 1024
INFO 05-27 11:35:58 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-27 11:35:59 [__init__.py:239] Automatically detected platform cuda.
pre-reorder-performance:
   batch_size    TopK=2    TopK=4    TopK=8
0        64.0  0.007552  0.008512  0.009824
1       128.0  0.007872  0.008352  0.010112
2       256.0  0.008224  0.008576  0.010048
3       512.0  0.008480  0.009344  0.011296
4       640.0  0.008704  0.009632  0.012224
5       768.0  0.008896  0.009792  0.012864
6      1024.0  0.009120  0.010816  0.014144
```

After fix, the result is:
```
[root@decfee1df170 fused_moe_triton]# python benchmark_pre_reorder_triton.py --hidden-size 1024
INFO 05-27 11:34:36 [importing.py:53] Triton module has been replaced with a placeholder.
INFO 05-27 11:34:37 [__init__.py:239] Automatically detected platform cuda.
pre-reorder-performance:
   batch_size    TopK=2    TopK=4    TopK=8
0        64.0  0.007360  0.008128  0.009600
1       128.0  0.007488  0.008224  0.009920
2       256.0  0.007840  0.008416  0.009920
3       512.0  0.008224  0.009088  0.011104
4       640.0  0.008480  0.009184  0.011808
5       768.0  0.008672  0.009632  0.012608
6      1024.0  0.008896  0.010592  0.014240
```

File: benchmark/kernels/fused_moe_triton/benchmark_pre_reorder_triton.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+import triton
+import triton.language as tl
+
+from sglang.srt.layers.moe.ep_moe.kernels import pre_reorder_triton_kernel
+
+
+@triton.testing.perf_report(
+    triton.testing.Benchmark(
+        x_names=["batch_size"],
+        x_vals=[64, 128, 256, 512, 640, 768, 1024],
```
Comment (BBuf) [inline benchmark/kernels/fused_moe_triton/benchmark_pre_reorder_triton.py]: The batch size (bs) and topk can be combined using the product function. Refer to https://github.com/sgl-project/sglang/blob/main/sgl-kernel/benchmark/bench_moe_align_block_size.py#L248 for more details.

Comment (yuan-luo) [reply benchmark/kernels/fused_moe_triton/benchmark_pre_reorder_triton.py]: Revised to use product.

**Review by BBuf — APPROVED** [review]
Please rename benchmark script to `benchmark_ep_pre_reorder_triton.py`

**Comment by yuan-luo** [comment]
> Please rename benchmark script to `benchmark_ep_pre_reorder_triton.py`

Done.

**Comment by BBuf** [comment]
Pr is ready now, please have a look when you are free? @zhyncs

**Review by Alcanderian — APPROVED** [review]
